omniledger/service/service.go

```diff
@@ -776,20 +811,168 @@ func (s *Service) getTxs(leader *network.ServerIdentity, scID skipchain.SkipBloc
 		return []ClientTransaction{}
 	}
 	if !leader.Equal(actualLeader) {
-		log.Lvl1("getTxs came from a wrong leader")
+		log.Warn(s.ServerIdentity(), "getTxs came from a wrong leader")
 		return []ClientTransaction{}
 	}
+	if s.heartbeats.enabled() {
+		s.heartbeats.beat(string(scID))
+	}
 	return s.txBuffer.take(string(scID))
 }
 
 // TestClose closes the go-routines that are polling for transactions.
 func (s *Service) TestClose() {
-	for _, c := range s.pollChan {
+	for k, c := range s.pollChan {
 		close(c)
+		delete(s.pollChan, k)
+	}
+	if s.heartbeats.enabled() {
+		s.heartbeats.closeAll()
+		s.heartbeatsClose <- true
 	}
 	s.pollChanWG.Wait()
 }
 
+func (s *Service) heartbeatTimeoutMonitor() {
```

Comment:
Maybe call it "monitorForLeaderFailure" or something. The heartbeat mechanism is how it gets done, but the reader of this code wants to know what is getting done.